### PR TITLE
Added PlaceholderText functionality to TextBox control

### PR DESCRIPTION
**Pull request by stefanov-stefan**
Added TextBox NullText support.

### CODE REVIEW

**Comment by AdamYoblick** [comment]
@stefanov-stefan Please close and reopen this PR to trigger the CI build again. Sorry for the inconvenience.

**Comment by filipnavara** [comment]
@dotnet-bot test this please (wanna see if the bot can trigger it on this repo)

**Comment by stefanov-stefan** [comment]
Will do

File: src/System.Windows.Forms/src/Resources/SR.resx [hunk 1]
```diff
@@ -7377,4 +7377,7 @@
   <data name="UnexpectedTypeForClipboardFormat" xml:space="preserve">
     <value>The specified Clipboard format is not compatible with '{0}' type.</value>
   </data>
-</root>
\ No newline at end of file
+  <data name="TextBoxNullTextDescr" xml:space="preserve">
+    <value>Specifies the NullText of the TextBox control.</value>
```
Comment (merriemcgaw) [inline src/System.Windows.Forms/src/Resources/SR.resx]: Can you please update this string to use "PlaceholderText"?

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/Resources/SR.resx]: It's not clear from this description, what the placeholder text is. Consider adding a second sentence explaining that this text is displayed when the Text property is null or empty and can be used as a suggestion or default value of the TextBox. 

---
In reply to: [239289457](https://github.com/dotnet/winforms/pull/96#discussion_r239289457) [](ancestors = 239289457)

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs [hunk 2]
```diff
@@ -855,9 +855,89 @@ internal void SetAutoComplete(bool reset)
                 }
             }
         }
-	
-	//-------------------------------------------------------------------------------------------------
-        
+
+        private string placeholderText;
+
+        /// <include file='doc\TextBox.uex' path='docs/doc[@for="TextBox.NullText"]/*' />
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: >  ///  [](start = 6, length = 6)

This comment style is outdated, could you please remove the include tag and replace devdoc tagwith summary - comments that are generated by the Visual Studio by default

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs [hunk 3]
```diff
@@ -855,9 +855,89 @@ internal void SetAutoComplete(bool reset)
                 }
             }
         }
-	
-	//-------------------------------------------------------------------------------------------------
-        
+
+        private string placeholderText;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: >private [](start = 8, length = 7)

In this file all private fields are placed in the beginning of the file, could you please move this one there as well? We would like to keep file style as consistent as possible.

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs [hunk 4]
```diff
@@ -855,9 +855,89 @@ internal void SetAutoComplete(bool reset)
...
+                    this.Invalidate();
+                }
+            }
+        }
+
+
+        //-------------------------------------------------------------------------------------------------
+
+        /// <summary>
+        /// Draws the NullText in the client area of the TextBox using the default font and color.
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: >NullText [](start = 22, length = 8)

please rename

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs [hunk 5]
```diff
@@ -855,9 +855,89 @@ internal void SetAutoComplete(bool reset)
...
+        /// </summary>
+        private void DrawPlaceholderText(Graphics graphics)
+        {
+            TextFormatFlags flags = TextFormatFlags.NoPadding | TextFormatFlags.Top |
+                                    TextFormatFlags.EndEllipsis;
+            Rectangle rectangle = this.ClientRectangle;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: >this [](start = 34, length = 4)

CoreFX style guide does not recommend `this.` unless required. See codding guidelines: [https://github.com/dotnet/corefx/tree/master/Documentation](https://github.com/dotnet/corefx/tree/master/Documentation)
We can't adopt all the guidelines, for example field name changes might affect binary serialization. But the rule about the `this` is less impactful and we'll update all sources at some point to follow it.

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs [hunk 6]
```diff
@@ -855,9 +855,89 @@ internal void SetAutoComplete(bool reset)
...
+
+
+        //-------------------------------------------------------------------------------------------------
+
+        /// <summary>
+        /// Draws the NullText in the client area of the TextBox using the default font and color.
+        /// </summary>
+        private void DrawPlaceholderText(Graphics graphics)
+        {
+            TextFormatFlags flags = TextFormatFlags.NoPadding | TextFormatFlags.Top |
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: >flags [](start = 28, length = 5)

Would it make sense to inherit alignment flags from the TextBox control?

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs [hunk 7]
```diff
@@ -855,9 +855,89 @@ internal void SetAutoComplete(bool reset)
...
+                flags |= TextFormatFlags.RightToLeft;
+                switch (this.TextAlign)
+                {
+                    case HorizontalAlignment.Center:
+                        flags = flags | TextFormatFlags.HorizontalCenter;
+                        rectangle.Offset(0, 1);
+                        break;
+                    case HorizontalAlignment.Left:
+                        flags = flags | TextFormatFlags.Right;
+                        rectangle.Offset(1, 1);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: offset values probably should be scaled for the current DPI

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs [hunk 8]
```diff
@@ -890,8 +970,31 @@ internal void SetAutoComplete(bool reset)
...
+                    DrawPlaceholderText(g);
+                }
+            }
+        }
+
+        protected override AccessibleObject CreateAccessibilityInstance()
+        {
+            if (string.IsNullOrEmpty(this.Text))
+            {
+                AccessibleObject accessibleObject = base.CreateAccessibilityInstance();
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: consider refactoring this code so that the base class method is called once.

**Comment by Tanya-Solyanik** [comment]
Had you checked ration for all high contrast modes?

---
In reply to: [443799542](https://github.com/dotnet/winforms/pull/96#issuecomment-443799542) [](ancestors = 443799542)

**Comment by jnm2** [comment]
@merriemcgaw @dotMorten What do you think of the name CueText, since that's what the Windows concept is named (https://docs.microsoft.com/windows/desktop/controls/em-setcuebanner)?

**Comment by dotMorten** [comment]
@jnm2 I hate it because we already set a precedence in the .NET world:
https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textbox.placeholdertext
https://docs.microsoft.com/en-us/dotnet/api/Xamarin.Forms.Entry.Placeholder?view=xamarin-forms

I've never even heard the concept "CueText" before. I'm going to guess that most people would not be able to guess what that property does. Being able to guess what a property does goes in hand with giving members good intuitive names.

**Comment by dotMorten** [comment]
Btw... looking at the CueBanner doc, I do like that it's possible to still show the placeholder while it has focus. Perhaps it would be a good idea to add a second property that allow you to show the placeholder text if it has focus (but text still being empty)

**Comment by jnm2** [comment]
@dotMorten Perfect. I wasn't aware of https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textbox.placeholdertext.

**Comment by nikeee** [comment]
I am really excited about this new feature!
One question, though: Why do you implemented the actual drawing? There have been other ways of achieving this that use the native Windows placeholder functionality. E.g.: https://github.com/nikeee/HolzShots/blob/master/src/HolzShots.UI/Windows/Forms/PromptTextBox.vb#L27 (using a SendMessage call).

**Comment by zsd4yr** [comment]
Yeah once #175 is done this should not be an issue

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs (hunk 8 above)
Comment (stefanov-stefan) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: the base class method is called only once - we will end either in the if or in the else in each case, we call it once

File: src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs (hunk 7 above)
Comment (stefanov-stefan) [reply src/System.Windows.Forms/src/System/Windows/Forms/TextBox.cs]: These values work well in high DPI scenario as well. I have tested it.

**Comment by stefanov-stefan** [comment]
sdf

**Comment by KlausLoeffelmann** [comment]
Hey Stefan,

Thanks for the screenshots!

Can you check, though, if those are rendered `DpiUnaware`? (Because it seems like that on first sight).
We would need screenshots for `SystemAware` and `PerMonitorAwareV2` with `VisualStyles` enabled.

Said that, it looks like the Window *is* `SystemAware`, but the content does not render correctly the with visual styles. Can you check if you got the latest bits for testing that? (We had a problem with resources not being properly included in *system.windows.forms.dll* in one of the previous versions - but that bug is fixed. If you have `EnableVisualStyles` in your *program.cs*, we probably got another problem I don't know of yet, and would need to investigate. In that case: Thanks for pointing out a bug! 😊).

**Comment by stefanov-stefan** [comment]
I attempted to run the CI with close and reopen, however when i closed git is not playing well with be and does not allow me to reopen. Can someone reopen this PR? @merriemcgaw @Tanya-Solyanik

**Comment by stefanov-stefan** [comment]
> Hey Stefan,
> 
> Thanks for the screenshots!
> 
> Can you check, though, if those are rendered `DpiUnaware`? (Because it seems like that on first sight).
> We would need screenshots for `SystemAware` and `PerMonitorAwareV2` with `VisualStyles` enabled.
> 
> Said that, it looks like the Window _is_ `SystemAware`, but the content does not render correctly then with visual styles. Can you check if you got the latest bits for testing that? (We had a problem with resources not being properly included in _system.windows.forms.dll_ in one of the previous versions which caused EnableVisualStyles to fail - but that bug is fixed. If you have `EnableVisualStyles` in your _program.cs_, we probably got another problem I don't know of yet, and would need to investigate. In that case: Thanks for pointing out a bug! 😊).

Yes, these shots are rendered with DpiUnaware. Attaching for `SystemAware` and `PerMonitorAwareV2`.
permonitorv2
aware

**Comment by KlausLoeffelmann** [comment]
Sorry, they are not. Also, VisualStyles are still off?
They should look something like this:

image

**Comment by stefanov-stefan** [comment]
VisualStyles are enabled as is by default. The app I am testing with is in the winforms sln WinformsControlsTests project. Here is the branch where I work: https://github.com/stefanov-stefan/winforms/tree/TextBox-NullText

**Comment by KlausLoeffelmann** [comment]
When you create a new WinForms Core App based on the latest build, and enable SystemAware, does it render correctly? Is your demo app VB or C# (there might be a problem with VB)?

**Comment by stefanov-stefan** [comment]
I am in FX app, not core with C#. Above I indicated the app I am using - it is the one included in the System.Windows.Forms.sln

**Comment by nikeee** [comment]
I'm curious, what's the advantage over using the SendMessage described above? Wouldn't this come with High-DPI support for free?

**Comment by jnm2** [comment]
@nikeee EM_SETCUEBANNER has no effect on multiline textboxes. This is a Windows constraint. The PR author really wants placeholder text with multiline.

I'm wondering how bad it would be to temporarily switch into single-line mode when text is null or empty and there is non-empty placeholder text. I think you have to recreate the underlying window with a new window style, so this could have a downside such as flicker or interruptions in behavior such as animations.

**Comment by AdamYoblick** [comment]
@stefanov-stefan I'm trying to reopen this PR and I'm not able to, even as a repo admin. :(

**Comment by karelz** [comment]
I think it is, because it was created when the repo was private. @raffaeler ran into the same thing and had to recreate his PR.

**Comment by raffaeler** [comment]
@AdamYoblick I can confirm. It was a pain because my repo lost the status of `forked` and I had to start from scratch (re-fork, clone the repo, diff the files whose state was of course different, etc.)

**Comment by stefanov-stefan** [comment]
I made it public, but it did not help. Will go over the process again and make new PR.

**Comment by bergmeister** [comment]
Since MSFT owns GitHub now, you should definitely feed this problem back to them and get it fixed.

**Comment by karelz** [comment]
@bergmeister I don't think that statement differs if MS owns GH or not. It is feedback which should be sent there.
How to prioritize is obviously up to the product group (GH in this case).
Once GH has bug reporting channel, we should try to report it there. Given that this is a rare case, it is not too high on my priority list (i.e. there is 20 other things I want GH to fix first as they have much higher impact on repos ;)).

**Comment by zsd4yr** [comment]
(fyi: I sent the feedback)

**Comment by stefanov-stefan** [comment]
I sent it too back when I experienced it, so I hope it will be taken into consideration.

**Comment by bergmeister** [comment]
Sorry for the late response and thank you for taking the time to send the feedback :)

**Comment by RussKie** [comment]
> One question, though: Why did you implement the actual drawing? There have been other ways of achieving this that use the native Windows placeholder functionality. E.g.: ...using a SendMessage call, sending the EM_SETCUEBANNER message.

This remains a big open question. Why have we opted to a custom implementation instead of leveraging the behaviours provided to us by the underlying Win32 control?

Not only the custom implementation introduced a number of issues, it is unaware of the native implementation and visually inconsistent with it:
```cs
            textBox1.PlaceholderText = "Custom!";
            SendMessageW(textBox1.Handle, EM_SETCUEBANNER, IntPtr.Zero, "NATIVE PlaceHolder..");
```
image

Note the visual differences:
* This implementation - `TextBox` control with `PlaceHolderText` property set:
image
* Native implementation - `TextBox` control with the placeholder text set via the `EM_SETCUEBANNER`:
image

    - different locations,
    - different font colours, and
    - different shortening strategies.


What's worse though the current implementation suffers from a quite an observable flickering when mouse is moved over the control:
* The top is `TextBox` control with `PlaceHolderText` property set, and
* the bottom is `TextBox` control with the placeholder text set via the `EM_SETCUEBANNER`
placeholdertext-existing


The new implementation is also not UIA aware:
* This implementation
image
* Native implementation:
image

**Comment by KlausLoeffelmann** [comment]
Can't we change this for 6? (And make it virtual along the way?)

**Pull request closed without merging**